### ISSUE TITLE
toco runs into flatbuffer 2GB limit when converting 181MB frozen graph_def

### ISSUE THREAD

**Issue by reuben** [tensorflow/tensorflow#21924]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: KDE Neon
- **TensorFlow installed from (source or binary)**: source
- **TensorFlow version (use command below)**: master branch, rev 57919740bf151cb6395aa60e30404ee9caa066d6
- **Python version**: 3.6.4
- **Bazel version (if compiling from source)**: 0.16.1
- **GCC/Compiler version (if compiling from source)**: gcc (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
- **CUDA/cuDNN version**: CPU-only build
- **GPU model and memory**: N/A
- **Mobile device**: N/A
- **Exact command to reproduce**: bazel run --compilation_mode=dbg tensorflow/contrib/lite/toco:toco -- --input_file=/path/to/output_graph.pb --inference_type=FLOAT --input_arrays=input_node --input_shapes=1,16,494 --output_array=logits --output_file=/tmp/graph.tflite

### Describe the problem

When trying to convert my GraphDef file to a TFLite file using TOCO I run into the 2GB limit for flatbuffers. The frozen graph protobuf is 181MB. I've uploaded the frozen graph protobuf as well as an unfrozen pbtxt here: https://github.com/reuben/silly_hacks/releases/tag/github_as_file_hosting_service

```
$ bazel run --compilation_mode=dbg tensorflow/contrib/lite/toco:toco -- --input_file=/home/reuben/output_graph.pb --inference_type=FLOAT --input_arrays=input_node --input_shapes=1,16,494 --output_array=logits --output_file=/home/reuben/graph.tflite
INFO: Analysed target //tensorflow/contrib/lite/toco:toco (0 packages loaded).
INFO: Found 1 target...
Target //tensorflow/contrib/lite/toco:toco up-to-date:
  bazel-bin/tensorflow/contrib/lite/toco/toco
INFO: Elapsed time: 0.213s, Critical Path: 0.00s
INFO: 0 processes.
INFO: Build completed successfully, 1 total action
INFO: Running command line: bazel-bin/tensorflow/contrib/lite/toco/toco '--input_file=/home/reuben/output_graph.pb' '--inference_type=FLOAT' '--input_arrays=input_node' '--input_shapes=1,16,494' '--output_INFO: Build completed successfully, 1 total action
2018-08-28 10:42:03.161524: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before Removing unused ops: 262 operators, 398 arrays (0 quantized)
2018-08-28 10:42:03.170762: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before general graph transformations: 262 operators, 398 arrays (0 quantized)
2018-08-28 10:42:11.384607: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 1: 225 operators, 373 arrays (0 quantized)
2018-08-28 10:42:11.454370: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 2: 223 operators, 371 arrays (0 quantized)
2018-08-28 10:42:11.472062: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] After general graph transformations pass 3: 221 operators, 367 arrays (0 quantized)
2018-08-28 10:42:11.487687: I tensorflow/contrib/lite/toco/graph_transformations/graph_transformations.cc:39] Before dequantization graph transformations: 221 operators, 367 arrays (0 quantized)
2018-08-28 10:42:11.493166: I tensorflow/contrib/lite/toco/allocate_transient_arrays.cc:345] Total transient array allocated size: 278528 bytes, theoretical optimal value: 262144 bytes.
2018-08-28 10:42:11.496319: I tensorflow/contrib/lite/toco/toco_tooling.cc:394] Estimated count of arithmetic ops: 1.52158 billion (note that a multiply-add is counted as 2 ops).
toco: external/flatbuffers/include/flatbuffers/flatbuffers.h:590: size_t flatbuffers::vector_downward::ensure_space(size_t): Assertion `size() < FLATBUFFERS_MAX_BUFFER_SIZE' failed.
Aborted
```

### Source code / logs

Traceback:

```
(gdb) bt
#0  0x00007ffff46a9428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff46ab02a in __GI_abort () at abort.c:89
#2  0x00007ffff46a1bd7 in __assert_fail_base (fmt=<optimized out>, assertion=assertion@entry=0x73cdf0 "size() < FLATBUFFERS_MAX_BUFFER_SIZE", file=file@entry=0x73cd28 "external/flatbuffers/include/flatbuffers/flatbuffers.h",
    line=line@entry=590, function=function@entry=0x73d640 <flatbuffers::vector_downward::ensure_space(unsigned long)::__PRETTY_FUNCTION__> "size_t flatbuffers::vector_downward::ensure_space(size_t)") at assert.c:92
#3  0x00007ffff46a1c82 in __GI___assert_fail (assertion=0x73cdf0 "size() < FLATBUFFERS_MAX_BUFFER_SIZE", file=0x73cd28 "external/flatbuffers/include/flatbuffers/flatbuffers.h", line=590,
    function=0x73d640 <flatbuffers::vector_downward::ensure_space(unsigned long)::__PRETTY_FUNCTION__> "size_t flatbuffers::vector_downward::ensure_space(size_t)") at assert.c:101
#4  0x00000000005baac8 in flatbuffers::vector_downward::ensure_space (this=0x7fffffffc8e0, len=0) at external/flatbuffers/include/flatbuffers/flatbuffers.h:590
#5  0x00000000005baaf1 in flatbuffers::vector_downward::make_space (this=0x7fffffffc8e0, len=0) at external/flatbuffers/include/flatbuffers/flatbuffers.h:595
#6  0x00000000005bacb9 in flatbuffers::vector_downward::fill (this=0x7fffffffc8e0, zero_pad_bytes=0) at external/flatbuffers/include/flatbuffers/flatbuffers.h:647
#7  0x00000000005bb12e in flatbuffers::FlatBufferBuilder::Align (this=0x7fffffffc8e0, elem_size=4) at external/flatbuffers/include/flatbuffers/flatbuffers.h:840
#8  0x00000000005bd1ad in flatbuffers::FlatBufferBuilder::PushElement<unsigned int> (this=0x7fffffffc8e0, element=134217728) at external/flatbuffers/include/flatbuffers/flatbuffers.h:861
#9  0x00000000005bb831 in flatbuffers::FlatBufferBuilder::EndVector (this=0x7fffffffc8e0, len=134217728) at external/flatbuffers/include/flatbuffers/flatbuffers.h:1148
#10 0x00000000005f6fce in flatbuffers::FlatBufferBuilder::CreateVector<unsigned char> (this=0x7fffffffc8e0,
    v=0x7fff645ea010 "Ƣ\256<\355\212#\274\v\261w<_\"\016\275(\371\254\275\314I\274\274\206\032\300<\232\037B=~8\236\274\213\366\036<\020\254\251<\314\363\360;2\262\360\274\345\252}=", len=134217728)
    at external/flatbuffers/include/flatbuffers/flatbuffers.h:1194
#11 0x00000000006af292 in toco::tflite::(anonymous namespace)::CopyBuffer<(toco::ArrayDataType)2> (array=..., builder=0x7fffffffc8e0) at tensorflow/contrib/lite/toco/tflite/types.cc:56
#12 0x00000000006aec36 in toco::tflite::DataBuffer::Serialize (array=..., builder=0x7fffffffc8e0) at tensorflow/contrib/lite/toco/tflite/types.cc:141
#13 0x00000000005b9a88 in toco::tflite::ExportBuffers (model=..., buffers_to_write=std::vector of length 368, capacity 512 = {...}, builder=0x7fffffffc8e0) at tensorflow/contrib/lite/toco/tflite/export.cc:307
#14 0x00000000005ba1b2 in toco::tflite::Export (model=..., allow_custom_ops=false, output_file_contents=0x7fffffffd070, ops_by_type=std::map with 80 elements = {...}) at tensorflow/contrib/lite/toco/tflite/export.cc:387
#15 0x00000000005b9b5c in toco::tflite::Export (model=..., allow_custom_ops=false, output_file_contents=0x7fffffffd070) at tensorflow/contrib/lite/toco/tflite/export.cc:317
#16 0x000000000052eae2 in toco::Export (toco_flags=..., model=..., allow_custom_ops=false, output_file_contents=0x7fffffffd070) at tensorflow/contrib/lite/toco/toco_tooling.cc:407
#17 0x00000000005135fa in toco::(anonymous namespace)::ToolMain (parsed_toco_flags=..., parsed_model_flags=...) at tensorflow/contrib/lite/toco/toco.cc:90
#18 0x00000000005138fe in main (argc=1, argv=0x7fffffffd9e8) at tensorflow/contrib/lite/toco/toco.cc:127
```

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
GPU model and memory
Mobile device

**Comment by reuben** [comment]
I've updated the issue.

**Comment by jdduke** [comment]
Thanks for the report @reuben . There was a similar issue reported internally that turned out to be related to some conflicting flatbuffer symbols (see also [this patch](https://github.com/tensorflow/tensorflow/commit/22e855159462b502dc3af138d254214bd02cf68b#diff-12251921ed1d6cc4b14367cfba47de82)). It's not clear if the issues are related, will take a look.

**Comment by jdduke** [comment]
Also, just to confirm, do you also hit this issue if you use `--compilation_mode=opt` instead of `--compilation_mode=dbg`?

**Comment by jdduke** [comment]
Reassigning as this looks like it could be an issue with LSTM conversion.

**Comment by lissyx** [comment]
Thanks @jdduke @miaout17, would you know how we might be able to help on that ?

**Comment by bjacob** [comment]
We are still investigating the best fixes internally, but as an immediate work-around, you can pass:

--dedupe_array_min_size_bytes=1000

this allows your command to complete and the resulting .tflite file size is 188939536 bytes.

**Comment by reuben** [comment]
Oh, neat, thanks! I'll give it a try.

**Comment by bjacob** [comment]
Fixes are going through integration at the moment so expect them soon in this repository. Once they land, the default behavior will be equivalent to --dedupe_array_min_size_bytes=64, and your command line should work as-is.

**Comment by miaout17** [comment]
The problem should be closed after commit 5330ede39fa2f1f7b3302bc316061baf180fab44. 
Do you mind to try this again? Thanks!

**Comment by lissyx** [comment]
Thanks, we'll keep you posted, but so far I know that @reuben was able to properly generate the model :)

**Comment by lissyx** [comment]
@miaout17 It seems that this issue is indeed properly fixed, but we'll likely have to open new ones :)
